### DISCUSSION TITLE
Implementing a QGraphicsView-like widget in SixtyFPS

### DISCUSSION THREAD

**Discussion by pmjobin** [slint-ui/slint#367] [Q&A] [answered]
Hello,

I am writing a Qt Widgets application which essentially consists of a layout editor for a physics simulation engine.

![Screenshot from 2021-07-30 02-20-10](https://user-images.githubusercontent.com/1387411/127607612-cceb0d13-e107-4460-9f70-2b90b37d23f9.png)

As you can see from the screenshot above, its main components are:

- A custom QGraphicsView-like widget (central panel),
- A property tree editor (right panel),
- A gallery of clickable thumbnails (left panel), and
- A script editor (bottom panel). 

I am considering porting this application to Rust and based on my cursory exploration of it, SixtyFPS appears to be the perfect candidate for my user interface needs. My main concern is with the QGraphicsView which can contain up to a few thousand items for complex scenes. What would be the most efficient way to render such scenes in SixtyFPS? Is it possible to draw inside a SixtyFPS window procedurally using the femtovg api directly? (In a way equivalent to overriding the paintEvent implementation of a QWidget...)

Also, regarding the property tree editor, is it possible to implement a QItemDelegate-like facility inside a ListView?

Thanks!
P-M

**Comment by ogoffart** [comment, marked as answer]
Hi, 

We still have a bit of work to do to support all these widgets, but we aim to be able to support such usecase.

For the graphics view, this can be done by using models drawing the shapes.
There is a similar example in 7gui/circledraw.60

The example in action: (this is only the UI part and does not have the native code that would do edits to the model)
 https://sixtyfps.io/snapshots/master/editor/?load_url=https://raw.githubusercontent.com/sixtyfpsui/sixtyfps/master/examples/7gui/circledraw.60   

> Is it possible to draw inside a SixtyFPS window procedurally using the femtovg api directly?

Not yet, but we should expose such API somehow.

> Also, regarding the property tree editor, is it possible to implement a QItemDelegate-like facility inside a ListView?

That is also not yet part of our standard widget set, but this can be done with a custom ListView.

**Reply by pmjobin** [reply]
Thank you for the quick and informative response. A detail I forgot to mention is that the central viewport camera can be zoomed in/out + moved around and individual shapes can be translated + rotated. Is there a method to apply an affine 2D transform to a child components like in QML?

**Reply by tronical** [reply]
At moment that's not exposed, but the underlying graphics libraries support affine 2D transformations.
